pkg/sentry/fs/proc/sys_net.go

```diff
@@ -280,11 +280,105 @@ func (p *proc) newSysNetCore(ctx context.Context, msrc *fs.MountSource, s inet.S
 	return newProcInode(ctx, d, msrc, fs.SpecialDirectory, nil)
 }
 
+// +stateify savable
+type ipForwarding struct {
+	stack   inet.Stack `state:"wait"`
+	enabled *bool
+	fsutil.SimpleFileInode
+}
+
+func newIPForwardingInode(ctx context.Context, msrc *fs.MountSource, s inet.Stack) *fs.Inode {
+	ts := &ipForwarding{
+		SimpleFileInode: *fsutil.NewSimpleFileInode(ctx, fs.RootOwner, fs.FilePermsFromMode(0444), linux.PROC_SUPER_MAGIC),
+		stack:           s,
+	}
+	sattr := fs.StableAttr{
+		DeviceID:  device.ProcDevice.DeviceID(),
+		InodeID:   device.ProcDevice.NextIno(),
+		BlockSize: usermem.PageSize,
+		Type:      fs.SpecialFile,
+	}
+	return fs.NewInode(ctx, ts, msrc, sattr)
+}
+
+// +stateify savable
+type ipForwardingFile struct {
+	fsutil.FileGenericSeek          `state:"nosave"`
+	fsutil.FileNoIoctl              `state:"nosave"`
+	fsutil.FileNoMMap               `state:"nosave"`
+	fsutil.FileNoSplice             `state:"nosave"`
+	fsutil.FileNoopRelease          `state:"nosave"`
+	fsutil.FileNoopFlush            `state:"nosave"`
+	fsutil.FileNoopFsync            `state:"nosave"`
+	fsutil.FileNotDirReaddir        `state:"nosave"`
+	fsutil.FileUseInodeUnstableAttr `state:"nosave"`
+	waiter.AlwaysReady              `state:"nosave"`
+
+	ipForwarding *ipForwarding
+}
+
+// GetFile implements fs.InodeOperations.GetFile.
+func (ipf *ipForwarding) GetFile(ctx context.Context, dirent *fs.Dirent, flags fs.FileFlags) (*fs.File, error) {
+	flags.Pread = true
+	flags.Pwrite = true
+	return fs.NewFile(ctx, dirent, flags, &ipForwardingFile{
+		ipForwarding: ipf,
+	}), nil
+}
+
+// Read implements fs.FileOperations.Read.
+func (f *ipForwardingFile) Read(ctx context.Context, _ *fs.File, dst usermem.IOSequence, offset int64) (int64, error) {
+	if offset != 0 {
+		return 0, io.EOF
+	}
+
+	if f.ipForwarding.enabled == nil {
+		ipf, err := f.ipForwarding.stack.IPForwardingEnabled()
+		if err != nil {
+			return 0, err
+		}
+		f.ipForwarding.enabled = &ipf
+	}
+
+	val := "0\n"
+	if *f.ipForwarding.enabled {
+		// Technically, this is not quite compatible with Linux. Linux
+		// stores these as an integer, so if you write "2" into
+		// ip_forward, you should get 2 back. Tough luck.
+		val = "1\n"
```

Comment:
Could you not store the 'enabled' value as an int in the ipForwarding struct?

Reply:
I haven't quite understand you. The ‘enabled’ is a bool value now at the ‘ipForwarding’ struct. I saved it in the struct to restore state of ip forwarding later when the kernel is restored from a file. Do you mean that restoring previous settings is not important? And I can drop the ‘enabled’ field from the ‘ipForwarding’ struct. BTW ‘tcpSack’ struct use this technique and I took it as an example.

Reply:
Sorry for the ambiguous wording. I was just thinking maybe you could store enabled not as a bool but as an int(32?) and write out the value of that int to be a bit more compatible.

save/restore is important so removing it altogether is not an option.

Reply:
fixed